cmake/pika_option.cmake

```diff
@@ -5,63 +5,82 @@
 # Distributed under the Boost Software License, Version 1.0. (See accompanying
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
+include(CMakeDependentOption)
 include(CMakeParseArguments)
 
 set(PIKA_OPTION_CATEGORIES "Generic" "Build Targets" "Thread Manager"
                            "Profiling" "Debugging" "Modules"
 )
 
-function(pika_option option type description default)
+macro(pika_option option type description default)
   set(options ADVANCED)
-  set(one_value_args CATEGORY MODULE)
+  set(one_value_args CATEGORY DEPENDS MODULE)
   set(multi_value_args STRINGS)
   cmake_parse_arguments(
     PIKA_OPTION "${options}" "${one_value_args}" "${multi_value_args}" ${ARGN}
   )
 
-  if(NOT DEFINED ${option})
-    set(${option}
-        ${default}
-        CACHE ${type} "${description}" FORCE
-    )
-    if(PIKA_OPTION_ADVANCED)
-      mark_as_advanced(${option})
+  if("${type}" STREQUAL "BOOL")
+    # Use regular CMake options for booleans
+    if(NOT PIKA_OPTION_DEPENDS)
+      option("${option}" "${description}" "${default}")
+    else()
+      cmake_dependent_option(
+        "${option}" "${description}" "${default}" "${PIKA_OPTION_DEPENDS}" OFF
```

Comment:
We set the default to `default`. The last `OFF` is the `<force>` variable from https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html?highlight=cmake_dependent_option#command:cmake_dependent_option:
```
Makes <option> available to the user if the [semicolon-separated list](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#cmake-language-lists) of conditions in <depends> are all true. Otherwise, a local variable named <option> is set to <force>.
```
i.e. it's the value that is set if the `PIKA_OPTION_DEPENDS` condition is not true, which in all cases we have so far should just turn off the option. In theory there could be options that want the inverse, but I think we can add that when we need it.